app/src/main/java/com/xmartlabs/template/controller/ServiceAndDatabaseController.java

```diff
@@ -0,0 +1,140 @@
+        .observeOn(Schedulers.io())
+        .scan((databaseEntities, serviceEntities) ->
+            databaseProvider.updateEntities(serviceEntities, conditions).blockingGet()
+        )
+        .observeOn(AndroidSchedulers.mainThread());
+  }
+
+  @CheckResult
+  @NonNull
+  protected Observable<D> getEntity(@NonNull Function<Id, Single<D>> serviceCall, Id id) {
+    if (id == null) {
+      return Observable.error(new IllegalStateException("Id cannot be null"));
+    }
+    return Observable.mergeDelayError(
+        databaseProvider.getEntity(id).toObservable(),
+        getServiceEntity(serviceCall, id).toObservable()
+    );
+  }
+
+  @CheckResult
+  @NonNull
+  protected Single<D> getEntityFromDbIfPresentOrGetFromService(@NonNull Function<Id, Single<D>> serviceCall, Id id) {
+    if (id == null) {
+      return Single.error(new IllegalStateException("Id cannot be null"));
+    }
+    return databaseProvider.getEntity(id)
+        .switchIfEmpty(getServiceEntity(serviceCall, id).toMaybe())
+        .toSingle();
+
+  }
+
+  @CheckResult
+  @NonNull
+  protected Single<D> getServiceEntity(@NonNull Function<Id, Single<D>> serviceCall, Id id) {
+    return makeIoCall(serviceCall.apply(id))
+        .doOnSuccess(databaseProvider::updateEntity);
+  }
+
+  @CheckResult
+  @NonNull
+  protected Single<D> createEntityAndGetValue(@NonNull Single<D> serviceCall) {
+    return makeIoCall(serviceCall)
+        .flatMap(databaseProvider::createEntity);
+  }
+
+  @CheckResult
+  @NonNull
+  protected Completable createEntity(@NonNull Single<D> serviceCall) {
+    return createEntityAndGetValue(serviceCall)
+        .toCompletable();
+  }
+
+  @CheckResult
+  @NonNull
+  protected Single<D> updateEntityAndGetValue(D entity, @NonNull BiFunction<Id, D, Completable> serviceCall) {
+    Id entityId = entity.getId();
+    if (entityId == null) {
+      return Single.error(new IllegalStateException("EntityId id cannot be null"));
+    }
+
+    return updateEntityAndGetValue(entity, serviceCall.apply(entity.getId(), entity));
+  }
+
+  @CheckResult
+  @NonNull
+  protected Single<D> updateEntityAndGetValue(D entity, Completable completable) {
+    return makeIoCall(completable)
+        .toSingleDefault(entity)
+        .flatMap(databaseProvider::updateEntity);
+  }
+
+  @CheckResult
+  @NonNull
+  protected Completable updateEntity(D entity, @NonNull BiFunction<Id, D, Completable> serviceCall) {
+    return updateEntityAndGetValue(entity, serviceCall)
+        .toCompletable();
+  }
+
+  @CheckResult
+  @NonNull
+  protected Completable deleteEntity(@NonNull Function<Id, Completable> serviceCall, @NonNull Id entityId) {
+    return makeIoCall(serviceCall.apply(entityId))
+        .concatWith(databaseProvider.deleteEntityFromId(entityId));
+  }
+
+  @CheckResult
+  @NonNull
+  protected Completable deleteEntity(@NonNull Function<Id, Completable> serviceCall, @NonNull D entity) {
+    Id entityId = entity.getId();
+    if (entityId == null) {
```

Comment:
@diegomedina248 is not saying `No`, to that. Just saying NO to the `if (entityId == null)` to be replaced by an Optional